design-rules.md

```diff
@@ -248,3 +248,95 @@
       public fun destroy_empty<Value>(wallet: &mut Wallet<Value>) {}
   }
   ```
+
+- **Keep your functions pure to maintain composability. Do not use `transfer::transfer` or `transfer::public_transfer` inside core functions.**
+
+  ```Move
+  module suitears::amm {
+      struct Pool has key {
+        id: UID
+      }
+
+      // ✅ Right
+      // Return the excess coins even if they have zero value.
+      public fun add_liquidity<CoinX, CoinY, LP_Coin>(pool: &mut Pool, coin_x: Coin<CoinX>, coin_y: Coin<CoinY>): (Coin<LpCoin>, Coin<CoinX>, Coin<CoinY>) {}
+
+      // ✅ Right
+      public fun entry_add_liquidity<CoinX, CoinY, LP_Coin>(pool: &mut Pool, coin_x: Coin<CoinX>, coin_y: Coin<CoinY>, ctx: &mut TxContext) {
+        let (lp_coin, coin_x, coin_y) = add_liquidity(pool, coin_x, coin_y);
+        transfer::public_transfer(lp_coin, tx_context::sender(ctx));
+        transfer::public_transfer(coin_x, tx_context::sender(ctx));
+        transfer::public_transfer(coin_y, tx_context::sender(ctx));
+      }
+
+      // ❌ Wrong
+      public fun add_liquidity<CoinX, CoinY, LP_Coin>(pool: &mut Pool, coin_x: Coin<CoinX>, coin_y: Coin<CoinY>, ctx: &mut TxContext): Coin<LpCoin> {
+        transfer::public_transfer(coin_x, tx_context::sender(ctx));
+        transfer::public_transfer(coin_y, tx_context::sender(ctx));
+
+        coin_lp
+      }
+  }
+  ```
+
+- **Do not pass mutable references of coins. Move the whole coin.**
```

Comment:
pass the Coin by value with the right amount directly, it's better for transaction readability from the frontend